sw/device/lib/crypto/drivers/hmac.c

```diff
@@ -397,17 +480,21 @@ status_t hmac_update(hmac_ctx_t *ctx, const uint8_t *data, size_t len) {
   // Keep writing incoming bytes
   msg_fifo_write(data, len - leftover_len);
 
+  // delay here, even  this small is enough
+  for (size_t i = 0; i < 40; i = launder32(i + 1))
+    ;
+
   // Time to tell HMAC HWIP to stop, because we do not have enough message
   // bytes for another round.
   uint32_t cmd_reg =
       bitfield_bit32_write(HMAC_CMD_REG_RESVAL, HMAC_CMD_HASH_STOP_BIT, 1);
   abs_mmio_write32(kHmacBaseAddr + HMAC_CMD_REG_OFFSET, cmd_reg);
 
-  // Wait for HMAC HWIP operation to be completed.
-  HARDENED_TRY(hmac_idle_wait());
-
-  // Store context into `ctx`.
-  context_save(ctx);
+  // Wait for HMAC HWIP operation to be completed and store context into `ctx`
+  // only if not already done in case of hanged HW.
+  if (hmac_idle_wait(ctx).value != OTCRYPTO_RECOV_ERR.value) {
+    context_save(ctx);
+  }
```

Comment:
This exact comparison will fail to trigger if `CRYPTO_STATUS_DEBUG` is set on the command line, since then every instance of `OTCRYPTO_RECOV_ERR` includes the line number on which the macro was called. It would be more robust to use `status_extract` and compare the resulting `code`:
https://github.com/lowRISC/opentitan/blob/da20a5887b2a48508cfc6334e8fa7377b40d7083/sw/device/lib/base/status.h#L221
It's just a tool for test debugging, so maybe not the most critical, but could make debugging harder. I'd be okay with not blocking the PR and creating an issue to fix this. There's more background on how cryptolib handles status codes here: https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crypto/impl/status.h